File: scripts/telegram-slogan-tile-assets.py
import argparse
import json
import math
import os
import shutil
import subprocess
import tempfile
from pathlib import Path

from PIL import Image

from telegram_slogan_tile_effects import effect_preview_slogans, render_slogan_tile_frame


SLOGANS = [
    {
        "key": "daily_update",
        "text": "DAILY UPDATE",
        "fallbackText": "\U0001F3AB DAILY UPDATE \U0001F3AB",
        "emoji": "\U0001F3AB",
        "tileCount": 6,
        "color": (249, 115, 22),
        "accent": (251, 146, 60),
        "effect": "marquee_text",
    }
]


def parse_args():
    parser = argparse.ArgumentParser(description="Generate KAITO Telegram slogan tile custom emoji videos.")
    parser.add_argument("--output", required=True)
    parser.add_argument("--manifest", default="")
    parser.add_argument("--preview", default="")
    parser.add_argument("--size", type=int, default=100)
    parser.add_argument("--duration", type=float, default=2.0)
    parser.add_argument("--fps", type=int, default=24)
    parser.add_argument("--crf", type=int, default=48)
    parser.add_argument("--ffmpeg", default="")
    return parser.parse_args()


def resolve_ffmpeg(explicit_path):
    if explicit_path:
        return explicit_path
    if os.environ.get("FFMPEG_PATH"):
        return os.environ["FFMPEG_PATH"]
    candidates = [
        Path.cwd() / "node_modules" / "@ffmpeg-installer" / "win32-x64" / "ffmpeg.exe",
        Path.cwd() / "node_modules" / "@ffmpeg-installer" / "linux-x64" / "ffmpeg",
        Path.cwd() / "node_modules" / "@ffmpeg-installer" / "darwin-x64" / "ffmpeg",
        Path.cwd() / "node_modules" / "@ffmpeg-installer" / "darwin-arm64" / "ffmpeg",
    ]
    for candidate in candidates:
        if candidate.exists():
            return str(candidate)
    found = shutil.which("ffmpeg")
    if found:
        return found
    raise RuntimeError("ffmpeg is required. Install npm dependencies or set FFMPEG_PATH.")



def render_slogan_frame(slogan, size, frame_index, total_frames):
    return render_slogan_tile_frame(slogan, size, frame_index, total_frames)


def encode_tile(frame_dir, output_path, ffmpeg_path, fps, crf):
    command = [
        ffmpeg_path,
        "-y",
        "-framerate", str(fps),
        "-i", str(frame_dir / "frame_%04d.png"),
        "-an",
        "-c:v", "libvpx-vp9",
        "-b:v", "0",
        "-crf", str(crf),
        "-pix_fmt", "yuva420p",
        "-auto-alt-ref", "0",
        str(output_path),
    ]
    result = subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
    if result.returncode != 0:
        raise RuntimeError(f"ffmpeg failed for {output_path.name}: {result.stderr.strip()}")


def render_slogan(slogan, output_dir, frame_root, ffmpeg_path, size, duration, fps, crf):
    total_frames = max(1, int(duration * fps))
    tile_dirs = []
    for index in range(slogan["tileCount"]):
        tile_dir = frame_root / f"{slogan['key']}_{index:02d}"
        tile_dir.mkdir(parents=True, exist_ok=True)
        tile_dirs.append(tile_dir)

    for frame_index in range(total_frames):
        frame = render_slogan_frame(slogan, size, frame_index, total_frames)
        for index, tile_dir in enumerate(tile_dirs):
            crop = frame.crop((index * size, 0, (index + 1) * size, size))
            crop.save(tile_dir / f"frame_{frame_index:04d}.png")

    tiles = []
    for index, tile_dir in enumerate(tile_dirs):
        key = f"{slogan['key']}_{index:02d}"
        file_name = f"{key}.webm"
        output_path = output_dir / file_name
        encode_tile(tile_dir, output_path, ffmpeg_path, fps, crf)
        tiles.append({
            "index": index,
            "key": key,
            "emoji": slogan["emoji"],
            "fileName": file_name,
            "outputPath": str(output_path),
        })
    return tiles


def render_preview(slogans, preview_path, size):
    padding = 10
    preview_slogans = effect_preview_slogans(slogans)
    width = max(slogan["tileCount"] * size for slogan in preview_slogans) + padding * 2
    height = len(preview_slogans) * (size + padding) + padding
    sheet = Image.new("RGBA", (width, height), (15, 23, 42, 255))
    for index, slogan in enumerate(preview_slogans):
        banner = render_slogan_frame(slogan, size, 0, 1)
        sheet.alpha_composite(banner, (padding, padding + index * (size + padding)))
    preview_path.parent.mkdir(parents=True, exist_ok=True)
    sheet.convert("RGB").save(preview_path)


def main():
    args = parse_args()
    output_dir = Path(args.output)
    output_dir.mkdir(parents=True, exist_ok=True)
    ffmpeg_path = resolve_ffmpeg(args.ffmpeg)

    files = []
    slogans = {}
    with tempfile.TemporaryDirectory(prefix="kaito-slogan-tile-frames-") as temp_root:
        frame_root = Path(temp_root)
        for slogan in SLOGANS:
            tiles = render_slogan(slogan, output_dir, frame_root, ffmpeg_path, args.size, args.duration, args.fps, args.crf)
            placeholder = "".join(tile["emoji"] for tile in tiles)
            slogans[slogan["key"]] = {
                "key": slogan["key"],
                "text": slogan["text"],
                "fallbackText": slogan["fallbackText"],
                "emoji": slogan["emoji"],
                "effect": slogan.get("effect", "premium_scan"),
                "placeholder": placeholder,
                "tiles": tiles,
            }
            for tile in tiles:
                files.append({
                    "sloganKey": slogan["key"],
                    **tile,
                })

    if args.preview:
        render_preview(SLOGANS, Path(args.preview), args.size)

    result = {
        "ok": True,
        "outputDir": str(output_dir),
        "previewPath": str(Path(args.preview)) if args.preview else "",
        "generated": len(files),
        "slogans": slogans,
        "files": files,
    }
    if args.manifest:
        manifest_path = Path(args.manifest)
        manifest_path.parent.mkdir(parents=True, exist_ok=True)
        manifest_path.write_text(json.dumps(result, indent=2), encoding="utf-8")
    print(json.dumps(result, indent=2))


if __name__ == "__main__":
    main()

File: scripts/telegram-slogan-tile-assets.js
import { mkdir, readFile } from 'node:fs/promises';
import { dirname, resolve } from 'node:path';
import { spawn } from 'node:child_process';
import { fileURLToPath } from 'node:url';

const DEFAULT_OUTPUT_DIR = resolve(process.cwd(), 'public', 'brand', 'slogan-tiles');
const DEFAULT_MANIFEST_PATH = resolve(process.cwd(), 'data', 'telegram-slogan-tile-assets.json');
const DEFAULT_PREVIEW_PATH = resolve(process.cwd(), 'public', 'brand', 'slogan-tile-preview.png');
const PYTHON_HELPER = resolve(process.cwd(), 'scripts', 'telegram-slogan-tile-assets.py');

export const SLOGAN_TILE_DEFINITIONS = [
  {
    key: 'daily_update',
    text: 'DAILY UPDATE',
    fallbackText: '🎫 DAILY UPDATE 🎫',
    emoji: '🎫',
    tileCount: 6,
    effect: 'marquee_text'
  }
];

export function buildSloganTilePythonArgs(options = {}) {
  const args = [
    PYTHON_HELPER,
    '--output', options.outputDir || DEFAULT_OUTPUT_DIR,
    '--manifest', options.manifestPath === false ? '' : (options.manifestPath || DEFAULT_MANIFEST_PATH),
    '--preview', options.previewPath === false ? '' : (options.previewPath || DEFAULT_PREVIEW_PATH),
    '--size', String(options.size || 100),
    '--duration', String(options.duration || 2),
    '--fps', String(options.fps || 24),
    '--crf', String(options.crf || 48)
  ];
  if (options.ffmpegPath) args.push('--ffmpeg', options.ffmpegPath);
  return args;
}

export async function generateSloganTileAssets(options = {}) {
  const outputDir = options.outputDir || DEFAULT_OUTPUT_DIR;
  const manifestPath = options.manifestPath === false ? '' : (options.manifestPath || DEFAULT_MANIFEST_PATH);
  const previewPath = options.previewPath === false ? '' : (options.previewPath || DEFAULT_PREVIEW_PATH);
  const runCommand = options.runCommand || runProcess;
  await mkdir(outputDir, { recursive: true });
  if (manifestPath) await mkdir(dirname(manifestPath), { recursive: true });
  if (previewPath) await mkdir(dirname(previewPath), { recursive: true });

  const pythonPath = options.pythonPath || process.env.PYTHON || 'python';
  const args = buildSloganTilePythonArgs({ ...options, outputDir, manifestPath, previewPath });
  await runCommand(pythonPath, args);

  if (!manifestPath) {
    const generated = SLOGAN_TILE_DEFINITIONS.reduce((sum, item) => sum + item.tileCount, 0);
    return { ok: true, outputDir, previewPath, generated };
  }
  return JSON.parse(await readFile(manifestPath, 'utf8'));
}

export function buildSloganTileEmojiMap({ assetManifest, uploadResult }) {
  const uploadByFile = new Map((uploadResult.stickers || []).map((sticker) => [sticker.fileName, sticker]));
  const slogans = {};
  for (const [key, slogan] of Object.entries(assetManifest.slogans || {})) {
    slogans[key] = {
      key,
      text: slogan.text,
      fallbackText: slogan.fallbackText,
      emoji: slogan.emoji,
      effect: slogan.effect || '',
      placeholder: slogan.placeholder,
      tiles: (slogan.tiles || []).map((tile) => {
        const uploaded = uploadByFile.get(tile.fileName) || {};
        return {
          ...tile,
          customEmojiId: uploaded.customEmojiId || tile.customEmojiId || ''
        };
      })
    };
  }

  return {
    packName: uploadResult.packName || '',
    title: uploadResult.title || 'KAITO AI SHOP Slogan Tiles',
    stickerType: 'custom_emoji',
    stickerFormat: uploadResult.stickerFormat || 'video',
    source: uploadResult.source || uploadResult.sourceDir || assetManifest.outputDir || '',
    generatedAt: uploadResult.generatedAt || new Date().toISOString(),
    customEmojiIdsByFile: Object.fromEntries(
      (uploadResult.stickers || []).map((sticker) => [sticker.fileName, sticker.customEmojiId]).filter(([, id]) => id)
    ),
    customEmojiIdsByBrand: Object.fromEntries(
      (uploadResult.stickers || []).map((sticker) => [sticker.brandKey, [sticker.customEmojiId]]).filter(([, ids]) => ids[0])
    ),
    slogans
  };
}

function runProcess(command, args) {
  return new Promise((resolvePromise, reject) => {
    const child = spawn(command, args, { stdio: ['ignore', 'pipe', 'pipe'] });
    let stderr = '';
    child.stderr.on('data', (chunk) => {
      stderr += chunk.toString();
    });
    child.on('error', reject);
    child.on('close', (code) => {
      if (code === 0) {
        resolvePromise();
        return;
      }
      reject(new Error(`${command} exited with code ${code}: ${stderr.trim()}`));
    });
  });
}

function parseArgs(argv) {
  const args = {
    outputDir: DEFAULT_OUTPUT_DIR,
    manifestPath: DEFAULT_MANIFEST_PATH,
    previewPath: DEFAULT_PREVIEW_PATH,
    size: 100,
    duration: 2,
    fps: 24,
    crf: 48
  };
  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    if (arg === '--output') args.outputDir = resolve(argv[++index]);
    else if (arg === '--manifest') args.manifestPath = argv[++index] ? resolve(argv[index]) : false;
    else if (arg === '--preview') args.previewPath = argv[++index] ? resolve(argv[index]) : false;
    else if (arg === '--size') args.size = Number(argv[++index]);
    else if (arg === '--duration') args.duration = Number(argv[++index]);
    else if (arg === '--fps') args.fps = Number(argv[++index]);
    else if (arg === '--crf') args.crf = Number(argv[++index]);
    else if (arg === '--ffmpeg') args.ffmpegPath = argv[++index];
    else if (arg === '--python') args.pythonPath = argv[++index];
    else if (arg === '--help' || arg === '-h') args.help = true;
    else throw new Error(`Unknown argument: ${arg}`);
  }
  return args;
}

function helpText() {
  return [
    'Usage:',
    '  npm.cmd run telegram:generate-slogan-tiles',
    '',
    'Options:',
    '  --output <dir>      Defaults to public/brand/slogan-tiles',
    '  --manifest <json>   Defaults to data/telegram-slogan-tile-assets.json',
    '  --preview <png>     Defaults to public/brand/slogan-tile-preview.png',
    '  --size <px>         Defaults to 100 for Telegram custom emoji video',
    '  --duration <sec>    Defaults to 2',
    '  --fps <fps>         Defaults to 24',
    '  --crf <value>       Defaults to 48',
    '  --ffmpeg <path>     Override ffmpeg binary path',
    '  --python <path>     Override Python executable'
  ].join('\n');
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  if (args.help) {
    console.log(helpText());
    return;
  }
  const result = await generateSloganTileAssets(args);
  console.log(JSON.stringify(result, null, 2));
}

if (process.argv[1] === fileURLToPath(import.meta.url)) {
  main().catch((error) => {
    console.error(error.message);
    process.exitCode = 1;
  });
}
